### PR TITLE
ref(platforms): Improve implementation of platformRegistry

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This change simplifies the `PlatformRegistry` in various ways. I will leave inline comments on the PR.

### CODE REVIEW

File: src/components/hooks/usePlatform.tsx [hunk 1]
```diff
@@ -34,62 +35,6 @@ const query = graphql`
   }
 `;
 
-export const formatCaseStyle = (style: string, value: string): string => {
```
Comment (evanpurkhiser) [inline src/components/hooks/usePlatform.tsx:37]: Didn't make sense for this to be exported in `usePlatforms`.

File: src/components/hooks/usePlatform.tsx [hunk 2]
```diff
@@ -34,62 +35,6 @@ const query = graphql`
...
-//   PascalCase,
-//   snake_case,
-// }
-
-// export enum SupportLevel {
-//   production,
-//   community,
-// }
-
-export type Guide = {
```
Comment (evanpurkhiser) [inline src/components/hooks/usePlatform.tsx:70]: Moved to a types file

File: src/components/guideGrid.tsx [hunk 3]
```diff
@@ -11,14 +11,14 @@ type Props = {
 
 export function GuideGrid({platform, className}: Props) {
   const [currentPlatform] = usePlatform(platform);
-  // platform might actually not be a platform, so lets handle that case gracefully
-  if (!(currentPlatform as Platform).guides) {
+
+  if (currentPlatform.type === 'guide') {
```
Comment (evanpurkhiser) [inline src/components/guideGrid.tsx:15]: We're now able to discriminate on the `type` property to indicate if it's a platform or a platform guide. This is a lot cleaner.

File: src/types/platform.tsx
Comment (evanpurkhiser) [inline src/types/platform.tsx:1]: Everything is typed much better in here, making a bit more clear the distinction between a `Platform` and a `PlatformGuide`.

File: src/types/platform.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+   */
+  key: string;
+  /**
+   * Same as key. Use `title` for a human readable platform name.
+   *
+   * @see Platform.key
+   */
+  name: string;
+  /**
+   * Used to discriminate a pltaform from a guide
```
Comment (gggritso) [inline src/types/platform.tsx:23]:
```suggestion
   * Used to discriminate a platform from a guide
```

File: src/types/platform.tsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * The name of a guide is the name of the guide on the file system
+   */
+  name: string;
+  /**
+   * The parent platform of the guide
+   */
+  platform: string;
+  /**
+   * Used to discriminate a pltaform from a guide
```
Comment (gggritso) [inline src/types/platform.tsx:52]:
```suggestion
   * Used to discriminate a platform from a guide
```

File: src/shared/platformRegistry.ts [hunk 6]
```diff
@@ -3,175 +3,279 @@
...
-const fillFallback = (config: Platform | Guide, results: (Platform | Guide)[]) => {
-  if (config.fallbackPlatform) {
-    const fallback = results.find(p => p.name === config.fallbackPlatform);
-    if (!fallback) {
-      throw new Error(`Unable to find fallbackPlatform: ${config.fallbackPlatform}`);
+  return guides;
+}
+
+/**
+ * Walks the platforms directory and procudes a list of {@link Platform}'s
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:158]:
```suggestion
 * Walks the platforms directory and produces a list of {@link Platform}'s
```

File: src/shared/platformRegistry.ts [hunk 7]
```diff
@@ -3,175 +3,279 @@
...
   'fallbackPlatform',
   'categories',
   'aliases',
-]);
+]) satisfies Set<keyof PlatformConfig>;
 
-const shareableConfig = new Set(['caseStyle', 'supportLevel', 'sdk', 'categories']);
+/**
+ * The PlatformConfig keys which will be automatically propagated. Used in a
+ * few propegation scenarios.
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:43]:
```suggestion
 * few propagation scenarios.
```

File: src/shared/platformRegistry.ts [hunk 8]
```diff
@@ -3,175 +3,279 @@
...
+  'categories',
+]) satisfies Set<keyof PlatformConfig>;
 
-const DEFAULTS = {
-  caseStyle: 'canonical',
-  supportLevel: 'production',
-};
+/**
+ * Extracts the frontmatter config object of a platform's index.mdx into an
+ * object gaurentted to only contain the keys defined in
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:59]:
```suggestion
 * object gauranteed to only contain the keys defined in
```

File: src/components/guideGrid.tsx [hunk 9]
```diff
@@ -11,14 +11,14 @@ type Props = {
 
 export function GuideGrid({platform, className}: Props) {
   const [currentPlatform] = usePlatform(platform);
```
Comment (gggritso) [inline src/components/guideGrid.tsx:13]: 🤔 a little confusing that `usePlatform` might return a guide, but I guess that's just the data model?

File: src/shared/platformRegistry.ts [hunk 10]
```diff
@@ -3,175 +3,279 @@
...
+import {
+  Platform,
+  PlatformCaseStyle,
+  PlatformConfig,
+  PlatformGuide,
+  PlatformSupportLevel,
+} from '../types';
+
+/**
+ * Common default values that all platform configs will recieve (but may be
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:18]:
```suggestion
 * Common default values that all platform configs will receive (but may be
```

File: src/shared/platformRegistry.ts [hunk 11]
```diff
@@ -3,175 +3,279 @@
...
+  return platforms;
+}
 
-  constructor(path = 'src/platforms') {
-    this.platforms = [];
-    this.path = path;
-    this._keyMap = {};
+/**
+ * When the given config object has a fallbackPlatform, this will ensure the
+ * `propegatedConfig` keys are present with their values from the
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:211]:
```suggestion
 * `propagatedConfig` keys are present with their values from the
```

File: src/shared/platformRegistry.ts [hunk 12]
```diff
@@ -3,175 +3,279 @@
...
+  platformGuideMapping: Record<string, Platform | PlatformGuide>;
+  /**
+   * The list of all available platforms. Platform guides are children of
+   * platforms which have them
+   */
+  platforms: Platform[];
+}
+
+/**
+ * Constructs the patlform registry mapping object.
```
Comment (gggritso) [inline src/shared/platformRegistry.ts:261]:
```suggestion
 * Constructs the platform registry mapping object.
```

File: src/types/platform.tsx [hunk 13]
```diff
@@ -0,0 +1,144 @@
+/**
+ * This is the canonical representation of a Platform within the docs.
+ *
+ * This object is resolve from the {@link PlatformConfig} type.
```
Comment (gggritso) [inline src/types/platform.tsx:4]:
```suggestion
 * This object is resolved from the {@link PlatformConfig} type.
```

**Review by gggritso — APPROVED** [review]
The data model is a little confusing here, but that's not really anything as far as this PR is concerned. LGTM, thanks! Just a buncha typos

**Review by dashed — APPROVED** [review]

**Review by shanamatthews — APPROVED** [review]
This hurt my brain a little, but I'm on board that this refactor helps

**Pull request merged**